extensions/standard-processors/processors/TailFile.cpp

```diff
@@ -653,29 +688,57 @@ void TailFile::onTrigger(const std::shared_ptr<core::ProcessContext> &, const st
   if (!session->existsFlowFileInRelationship(Success)) {
     yield();
   }
+
+  first_trigger_ = false;
+}
+
+bool TailFile::isOldFileInitiallyRead(TailState &state) const {
+  // This is our initial processing and no stored state was found
+  return first_trigger_ && state.last_read_time_ == std::chrono::system_clock::time_point{};
 }
 
 void TailFile::processFile(const std::shared_ptr<core::ProcessSession> &session,
                            const std::string &full_file_name,
                            TailState &state) {
-  uint64_t fsize = utils::file::FileUtils::file_size(full_file_name);
-  if (fsize < state.position_) {
-    processRotatedFiles(session, state);
-  } else if (fsize == state.position_) {
-    logger_->log_trace("Skipping file %s as its size hasn't change since last read", state.file_name_);
-    return;
+  if (isOldFileInitiallyRead(state)) {
+    if (initial_start_position_ == InitialStartPositions::BEGINNING_OF_TIME) {
+      processAllRotatedFiles(session, state);
+    } else if (initial_start_position_ == InitialStartPositions::CURRENT_TIME) {
+      state.position_ = utils::file::FileUtils::file_size(full_file_name);
+      state.last_read_time_ = std::chrono::system_clock::now();
+      state.checksum_ = utils::file::FileUtils::computeChecksum(full_file_name, state.position_);
+      storeState();
+      return;
+    }
+  } else {
+    uint64_t fsize = utils::file::FileUtils::file_size(full_file_name);
+    if (fsize < state.position_) {
+      processRotatedFilesAfterLastReadTime(session, state);
+    } else if (fsize == state.position_) {
+      logger_->log_trace("Skipping file %s as its size hasn't change since last read", state.file_name_);
+      return;
+    }
   }
 
   processSingleFile(session, full_file_name, state);
 }
 
-void TailFile::processRotatedFiles(const std::shared_ptr<core::ProcessSession> &session, TailState &state) {
-    std::vector<TailState> rotated_file_states = findRotatedFiles(state);
-    for (TailState &file_state : rotated_file_states) {
-      processSingleFile(session, file_state.fileNameWithPath(), file_state);
-    }
-    state.position_ = 0;
-    state.checksum_ = 0;
+void TailFile::processRotatedFilesAfterLastReadTime(const std::shared_ptr<core::ProcessSession> &session, TailState &state) {
+  std::vector<TailState> rotated_file_states = findRotatedFilesAfterLastReadTime(state);
+  processRotatedFiles(session, state, rotated_file_states);
+}
+
+void TailFile::processAllRotatedFiles(const std::shared_ptr<core::ProcessSession> &session, TailState &state) {
+  std::vector<TailState> rotated_file_states = findAllRotatedFiles(state);
+  processRotatedFiles(session, state, rotated_file_states);
+}
+
+void TailFile::processRotatedFiles(const std::shared_ptr<core::ProcessSession> &session, TailState &state, std::vector<TailState> &rotated_file_states) {
+  for (TailState &file_state : rotated_file_states) {
+    processSingleFile(session, file_state.fileNameWithPath(), file_state);
```

Comment:
`processSingleFile` calls `storeState`, is that desirable with rotated files?

Reply:
It should be for a restart scenario, when the agent is restarted we should know that those files are already processed and not new files, as in that case we do not check for rollover patterns.

Reply:
I ment that `storeState` persists the `tail_states_` vector, which, as fas as I understand, does not contains the `state` parameter when called with rotated files, so in this case `storeState` seems noop (or rather "misleading"), could moving `storeState` to the end of `processFile` improve it in this regard?

Reply:
Thanks, now I see what you mean. I suppose you are right and in that case, we do not need that call for those temporary states anyway, it may be more clearer just to call after the `processSingleFile` call of the actual main file. Updated in 18d55b56cf8c4afc083fdb3ba60bc5c92985af27